File: output/go/tests/mealy_machine.go

package main

import "fmt"

// Mealy Machine - output depends on state AND input (output on transitions)
// See: https://www.geeksforgeeks.org/mealy-and-moore-machines-in-toc/

type MealyMachineFrameEvent struct {
    _message    string
    _parameters map[string]any
}

type MealyMachineFrameContext struct {
    _event  MealyMachineFrameEvent
    _return any
    _data   map[string]any
}

type MealyMachineCompartment struct {
    state            string
    stateArgs        map[string]any
    stateVars        map[string]any
    enterArgs        map[string]any
    exitArgs         map[string]any
    forwardEvent     *MealyMachineFrameEvent
    parentCompartment *MealyMachineCompartment
}

func newMealyMachineCompartment(state string) *MealyMachineCompartment {
    return &MealyMachineCompartment{
        state:    state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
    }
}

func (c *MealyMachineCompartment) copy() *MealyMachineCompartment {
    nc := &MealyMachineCompartment{
        state: c.state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
        forwardEvent:     c.forwardEvent,
        parentCompartment: c.parentCompartment,
    }
    for k, v := range c.stateArgs { nc.stateArgs[k] = v }
    for k, v := range c.stateVars { nc.stateVars[k] = v }
    for k, v := range c.enterArgs { nc.enterArgs[k] = v }
    for k, v := range c.exitArgs { nc.exitArgs[k] = v }
    return nc
}

type MealyMachine struct {
    _state_stack []*MealyMachineCompartment
    __compartment *MealyMachineCompartment
    __next_compartment *MealyMachineCompartment
    _context_stack []MealyMachineFrameContext
    last_output int
}

func NewMealyMachine() *MealyMachine {
    s := &MealyMachine{}
    s._state_stack = make([]*MealyMachineCompartment, 0)
    s._context_stack = make([]MealyMachineFrameContext, 0)
    s.__compartment = newMealyMachineCompartment("Q0")
    s.__next_compartment = nil
    __frame_event := MealyMachineFrameEvent{_message: "$>", _parameters: nil}
    __ctx := MealyMachineFrameContext{_event: __frame_event, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return s
}

func (s *MealyMachine) __kernel(__e *MealyMachineFrameEvent) {
    s.__router(__e)
    for s.__next_compartment != nil {
        next_compartment := s.__next_compartment
        s.__next_compartment = nil
        exit_event := &MealyMachineFrameEvent{_message: "<$", _parameters: s.__compartment.exitArgs}
        s.__router(exit_event)
        s.__compartment = next_compartment
        if s.__compartment.forwardEvent == nil {
            enter_event := &MealyMachineFrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
            s.__router(enter_event)
        } else {
            forward_event := s.__compartment.forwardEvent
            s.__compartment.forwardEvent = nil
            if forward_event._message == "$>" {
                s.__router(forward_event)
            } else {
                enter_event := &MealyMachineFrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
                s.__router(enter_event)
                s.__router(forward_event)
            }
        }
    }
}

func (s *MealyMachine) __router(__e *MealyMachineFrameEvent) {
    switch s.__compartment.state {
    case "Q0":
        s._state_Q0(__e)
    case "Q1":
        s._state_Q1(__e)
    case "Q2":
        s._state_Q2(__e)
    }
}

func (s *MealyMachine) __transition(next *MealyMachineCompartment) {
    s.__next_compartment = next
}

func (s *MealyMachine) I0() {
    __e := MealyMachineFrameEvent{_message: "I0"}
    __ctx := MealyMachineFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
}

func (s *MealyMachine) I1() {
    __e := MealyMachineFrameEvent{_message: "I1"}
    __ctx := MealyMachineFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
}

func (s *MealyMachine) _state_Q0(__e *MealyMachineFrameEvent) {
    if __e._message == "I0" {
        s.emitOutput(0)
        __compartment := newMealyMachineCompartment("Q1")
        __compartment.parentCompartment = s.__compartment.copy()
        s.__transition(__compartment)
        return
    } else if __e._message == "I1" {
        s.emitOutput(0)
        __compartment := newMealyMachineCompartment("Q2")
        __compartment.parentCompartment = s.__compartment.copy()
        s.__transition(__compartment)
        return
    }
}

func (s *MealyMachine) _state_Q2(__e *MealyMachineFrameEvent) {
    if __e._message == "I0" {
        s.emitOutput(1)
        __compartment := newMealyMachineCompartment("Q1")
        __compartment.parentCompartment = s.__compartment.copy()
        s.__transition(__compartment)
        return
    } else if __e._message == "I1" {
        s.emitOutput(0)
        __compartment := newMealyMachineCompartment("Q2")
        __compartment.parentCompartment = s.__compartment.copy()
        s.__transition(__compartment)
        return
    }
}

func (s *MealyMachine) _state_Q1(__e *MealyMachineFrameEvent) {
    if __e._message == "I0" {
        s.emitOutput(0)
        __compartment := newMealyMachineCompartment("Q1")
        __compartment.parentCompartment = s.__compartment.copy()
        s.__transition(__compartment)
        return
    } else if __e._message == "I1" {
        s.emitOutput(1)
        __compartment := newMealyMachineCompartment("Q2")
        __compartment.parentCompartment = s.__compartment.copy()
        s.__transition(__compartment)
        return
    }
}

func (s *MealyMachine) emitOutput(value int) {
                s.last_output = value
}

func (s *MealyMachine) GetLastOutput() int {
                return s.last_output
}

func main() {
	fmt.Println("TAP version 14")
	fmt.Println("1..4")

	m := NewMealyMachine()

	m.I0() // Q0 -> Q1, output 0
	if m.GetLastOutput() == 0 {
		fmt.Println("ok 1 - mealy i_0 from Q0 outputs 0")
	} else {
		fmt.Printf("not ok 1 - mealy i_0 from Q0 outputs 0 # got %v\n", m.GetLastOutput())
	}

	m.I0() // Q1 -> Q1, output 0
	if m.GetLastOutput() == 0 {
		fmt.Println("ok 2 - mealy i_0 from Q1 outputs 0")
	} else {
		fmt.Printf("not ok 2 - mealy i_0 from Q1 outputs 0 # got %v\n", m.GetLastOutput())
	}

	m.I1() // Q1 -> Q2, output 1
	if m.GetLastOutput() == 1 {
		fmt.Println("ok 3 - mealy i_1 from Q1 outputs 1")
	} else {
		fmt.Printf("not ok 3 - mealy i_1 from Q1 outputs 1 # got %v\n", m.GetLastOutput())
	}

	m.I0() // Q2 -> Q1, output 1
	if m.GetLastOutput() == 1 {
		fmt.Println("ok 4 - mealy i_0 from Q2 outputs 1")
	} else {
		fmt.Printf("not ok 4 - mealy i_0 from Q2 outputs 1 # got %v\n", m.GetLastOutput())
	}

	fmt.Println("# PASS - Mealy machine outputs depend on state AND input")
}

File: output/javascript/tests/mealy_machine.mjs

// Mealy Machine - output depends on state AND input (output on transitions)
// See: https://www.geeksforgeeks.org/mealy-and-moore-machines-in-toc/

export class MealyMachineFrameEvent {
    _message;
    _parameters;

    constructor(message, parameters) {
        this._message = message;
        this._parameters = parameters;
    }
}


export class MealyMachineFrameContext {
    event;
    _return;
    _data;

    constructor(event, default_return) {
        this.event = event;
        this._return = default_return;
        this._data = {  };
    }
}


export class MealyMachineCompartment {
    state;
    state_args;
    state_vars;
    enter_args;
    exit_args;
    forward_event;
    parent_compartment;

    constructor(state, parent_compartment = null) {
        this.state = state;
        this.state_args = {  };
        this.state_vars = {  };
        this.enter_args = {  };
        this.exit_args = {  };
        this.forward_event = null;
        this.parent_compartment = parent_compartment;
    }

    copy() {
        const c = new MealyMachineCompartment(this.state, this.parent_compartment);
        c.state_args = {...this.state_args};
        c.state_vars = {...this.state_vars};
        c.enter_args = {...this.enter_args};
        c.exit_args = {...this.exit_args};
        c.forward_event = this.forward_event;
        return c;
    }
}


export class MealyMachine {
    _state_stack;
    __compartment;
    __next_compartment;
    _context_stack;
    last_output = -1;

    constructor() {
        this._state_stack = [];
        this._context_stack = [];
        this.__compartment = new MealyMachineCompartment("Q0");
        this.__next_compartment = null;
        const __frame_event = new MealyMachineFrameEvent("$>", null);
        this.__kernel(__frame_event);
    }

    __kernel(__e) {
        // Route event to current state
        this.__router(__e);
        // Process any pending transition
        while (this.__next_compartment !== null) {
            const next_compartment = this.__next_compartment;
            this.__next_compartment = null;
            // Exit current state
            const exit_event = new MealyMachineFrameEvent("<$", this.__compartment.exit_args);
            this.__router(exit_event);
            // Switch to new compartment
            this.__compartment = next_compartment;
            // Enter new state (or forward event)
            if (next_compartment.forward_event === null) {
                const enter_event = new MealyMachineFrameEvent("$>", this.__compartment.enter_args);
                this.__router(enter_event);
            } else {
                // Forward event to new state
                const forward_event = next_compartment.forward_event;
                next_compartment.forward_event = null;
                if (forward_event._message === "$>") {
                    // Forwarding enter event - just send it
                    this.__router(forward_event);
                } else {
                    // Forwarding other event - send $> first, then forward
                    const enter_event = new MealyMachineFrameEvent("$>", this.__compartment.enter_args);
                    this.__router(enter_event);
                    this.__router(forward_event);
                }
            }
        }
    }

    __router(__e) {
        const state_name = this.__compartment.state;
        const handler_name = `_state_${state_name}`;
        const handler = this[handler_name];
        if (handler) {
            handler.call(this, __e);
        }
    }

    __transition(next_compartment) {
        this.__next_compartment = next_compartment;
    }

    i_0() {
        const __e = new MealyMachineFrameEvent("i_0", null);
        const __ctx = new MealyMachineFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        this._context_stack.pop();
    }

    i_1() {
        const __e = new MealyMachineFrameEvent("i_1", null);
        const __ctx = new MealyMachineFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        this._context_stack.pop();
    }

    _state_Q0(__e) {
        if (__e._message === "i_0") {
            this.emit_output(0);
            const __compartment = new MealyMachineCompartment("Q1", this.__compartment.copy());
            this.__transition(__compartment);
            return;
        } else if (__e._message === "i_1") {
            this.emit_output(0);
            const __compartment = new MealyMachineCompartment("Q2", this.__compartment.copy());
            this.__transition(__compartment);
            return;
        }
    }

    _state_Q1(__e) {
        if (__e._message === "i_0") {
            this.emit_output(0);
            const __compartment = new MealyMachineCompartment("Q1", this.__compartment.copy());
            this.__transition(__compartment);
            return;
        } else if (__e._message === "i_1") {
            this.emit_output(1);
            const __compartment = new MealyMachineCompartment("Q2", this.__compartment.copy());
            this.__transition(__compartment);
            return;
        }
    }

    _state_Q2(__e) {
        if (__e._message === "i_0") {
            this.emit_output(1);
            const __compartment = new MealyMachineCompartment("Q1", this.__compartment.copy());
            this.__transition(__compartment);
            return;
        } else if (__e._message === "i_1") {
            this.emit_output(0);
            const __compartment = new MealyMachineCompartment("Q2", this.__compartment.copy());
            this.__transition(__compartment);
            return;
        }
    }

    emit_output(value) {
                    this.last_output = value;
    }

    get_last_output() {
                    return this.last_output;
    }
}

function main() {
    console.log("TAP version 14");
    console.log("1..4");

    const m = new MealyMachine();

    m.i_0();  // Q0 -> Q1, output 0
    if (m.get_last_output() === 0) {
        console.log("ok 1 - mealy i_0 from Q0 outputs 0");
    } else {
        console.log(`not ok 1 - mealy i_0 from Q0 outputs 0 # got ${m.get_last_output()}`);
    }

    m.i_0();  // Q1 -> Q1, output 0
    if (m.get_last_output() === 0) {
        console.log("ok 2 - mealy i_0 from Q1 outputs 0");
    } else {
        console.log(`not ok 2 - mealy i_0 from Q1 outputs 0 # got ${m.get_last_output()}`);
    }

    m.i_1();  // Q1 -> Q2, output 1
    if (m.get_last_output() === 1) {
        console.log("ok 3 - mealy i_1 from Q1 outputs 1");
    } else {
        console.log(`not ok 3 - mealy i_1 from Q1 outputs 1 # got ${m.get_last_output()}`);
    }

    m.i_0();  // Q2 -> Q1, output 1
    if (m.get_last_output() === 1) {
        console.log("ok 4 - mealy i_0 from Q2 outputs 1");
    } else {
        console.log(`not ok 4 - mealy i_0 from Q2 outputs 1 # got ${m.get_last_output()}`);
    }

    console.log("# PASS - Mealy machine outputs depend on state AND input");
}

main();
